### DISCUSSION TITLE
USER STORY DRAFT: batch lockup creation

### DISCUSSION THREAD

**Discussion by alexaunter** [galactic3/ft-lockup-contract#32] [Q&A]
As DAO / multisig account, I can create 10_000 lockups
without signing/confirming each individual transaction, while not compromising
security of the contract.

Acceptance criteria:
- account holding tokens to be distributed doesn't have any full access keys and can't have them even for the creation of lockups
- it must be possible to confirm any number of lockup creations with constant number of transactions issued by token owner account

To be discussed.

**Comment by alexaunter** [comment]
Possible solution:
proxy contract is created.
proxy contract is filled with data about all future lockups to be created
all access keys are revoked from the proxy contract
DAO/multisig votes for creation of function access key that can call any methods of proxy account on behalf of token holder account, AND for transfer of total amount of tokens to be transferred to the proxy contract
Request is approved/confirmed, key is created.
Function call access key is used to call methods on proxy contract, proxy contract then calls add lockup on lockup contract without necessity to vote/confirm transaction.

**Reply by alexaunter** [reply]
FYI @AngelBlock, @frol

**Reply by AngelBlock** [reply]
Just to clarify, the proxy contract will be funded from a DAO or any other way (multisig, ledger or normal account), not that important how; and will hold tokens until they distributed. There should be a mechanism to allow for refunds to original sender / funder.  This will allow a batch creation, but can also be utilised for single distributions, if no way found to make both working same time.

So in fact DAO will not approve the distribution. but instead transfer the funds to that contract.

**Reply by alexaunter** [reply]
Another solution might be to merge this confirmation proxy functionality into the ft-lockup contract itself.
Then again, DAO will only fund the ft lockup contract one time, and then function key can be used to convert "pre-created" lockup into active, funded lockup.

**Comment by frol** [comment]
Just to reiterate the decision I made during the call is that we will go with the solution @alexaunter proposed in the https://github.com/galactic3/ft-lockup-contract/discussions/32#discussioncomment-2206589:

> Another solution might be to merge this confirmation proxy functionality into the ft-lockup contract itself.
Then again, DAO will only fund the ft lockup contract one time, and then function key can be used to convert "pre-created" lockup into active, funded lockup.
